---
title: "Single Cell SCLC Human Tumor Archetypes"
output:
  html_document:
    df_print: paged
---
The purpose of the notebook is to apply PCHA to the single cell cell line data. 

```{r include = FALSE}
library(reticulate)
reticulate::use_condaenv("/Users/smgroves/Documents/anaconda3/envs/mazebox_env", conda = "auto", required = TRUE) # set TRUE to force R to use reticulate_PCHA
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

# BiocManager::install("sva")
# install.packages("devtools")
# devtools::install_local("/Users/smgroves/Downloads/CytoTRACE-master.zip")
# library(CytoTRACE)
# BiocManager::install("scater")
library(scater)
# install loomR from GitHub using the remotes package 
# remotes::install_github(repo ='mojaveazure/loomR', ref = 'develop')
library(loomR)
# install.packages('Seurat')
library(Seurat)
# remotes::install_github("mojaveazure/seurat-disk")
# library(SeuratData)
library(SeuratDisk)


```

# Read single cell data 
Reading in preprocessed (MAGIC imputed) data from from `2-Human_cell_lines_scPCHA.ipynb`

```{r message = FALSE}
file = "../../out/human-tumors/X_magic_for_R.h5ad"
Convert(file, dest = "h5seurat", overwrite = TRUE)
data <- LoadH5Seurat("../../out/human-tumors/X_magic_for_R.h5seurat")
Idents(data) = data@meta.data$tumor

ElbowPlot(data) #this plot indicates  the top 8 are important
DimPlot(data, reduction = "pca", label = TRUE)
pct <- data[["pca"]]@stdev / sum(data[["pca"]]@stdev) * 100

# Calculate cumulative percents for each PC
cumu <- cumsum(pct)

# Determine which PC exhibits cumulative percent greater than 90% and % variation associated with the PC is less than 5
co1 <- which(cumu > 90 & pct < 5)[1]

print(co1) #37
# Determine the difference between variation of PC and subsequent PC
co2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1), decreasing = T)[1] + 1

# last point where change of % of variation is more than 0.1%.
print(co2) #12

data <- FindNeighbors(data, dims = 1:10)
data <- FindClusters(data, resolution = 0.5)
data <- RunUMAP(data, dims = 1:10)
DimPlot(data, reduction = "umap")

```


# Plot important SCLC Genes and add tumor labels

```{r}
data[["louvain"]] <- Idents(object = data) #store louvain Idents
Idents(object = data) <- data@meta.data$tumor #add cell line labels

DimPlot(data, label = TRUE) 
VlnPlot(data, features = c("ASCL1", "YAP1", "POU2F3", 'CALCA','NEUROD2', 'MYC'), combine = FALSE)
```

# Plotting the data

```{r include=FALSE}
# 
# library(dplyr)
# data.markers <- FindAllMarkers(data, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
# data.markers.top <- data.markers %>% group_by(cluster) %>% top_n(n = 30, wt = avg_log2FC)
```

```{r pca-umap-plots}

# install.packages('plotly')

# Load plot_ly
library(plotly)

# Extract tSNE information from Seurat Object
pca_1 <- data[["pca"]]@cell.embeddings[,1]
pca_2 <- data[["pca"]]@cell.embeddings[,2]
pca_3 <- data[["pca"]]@cell.embeddings[,3]

# Visualize what headings are called so that you can extract them to form a dataframe
# Embeddings(object = data, reduction = "pca")
plot.data <- FetchData(object = data, vars = c("PC_1", "PC_2", "PC_3", "tumor"))
plot.data$label <- paste(rownames(plot.data))
plot_ly(data = plot.data, 
        x = ~PC_1, y = ~PC_2, z = ~PC_3, 
        color = ~tumor, 
        colors = c("lightseagreen",
                   "green",
                   "red",
                   "orange1",
                   "royalblue1",
                   "lightcyan3",
                   "peachpuff3",
                   "darkorchid1",
                   "turquoise",
                   "darkmagenta"),
        type = "scatter3d", 
        mode = "markers", 
        marker = list(size = 5, width=2), # controls size of points
        text=~label, #This is that extra column we made earlier for which we will use for cell ID
        hoverinfo="text") #When you visualize your plotly object, hovering your mouse pointer over a point shows cell names

VizDimLoadings(data, dims = 1:3, reduction = "pca")

```

## Save files
```{r include = FALSE}

save(data, file="../../out/human-tumors/ParetoTI/X_magic_for_ParetoTI.Robj")

write.csv(data[["RNA"]]@scale.data, '../../out/human-tumors/ParetoTI/integrated-corrected-data.csv')
write.csv(data[['pca']]@cell.embeddings, '../../out/human-tumors/ParetoTI/pca_embedding.csv')
write.csv(data[['pca']]@feature.loadings, '../../out/human-tumors/ParetoTI/pca_feature_loadings.csv')

data <- ProjectDim(data, reduction = 'pca')
write.csv(data[['pca']]@feature.loadings.projected, '../../out/human-tumors/ParetoTI/pca_feature_loadings_projected.csv')

# devtools::install_github(repo = 'hhoeflin/hdf5r')
# devtools::install_github(repo = 'mojaveazure/loomR', ref = 'develop')
# devtools::install_github(repo = 'satijalab/seurat', ref = 'loom')
# library(loomR)
# 
# data.loom <- as.loom(data, filename = "X_magic_for_ParetoTI.loom", verbose = FALSE)
```

# Archetypal Analysis

We will now apply the archetype analysis to the single cell data. Since we've reduced the PCs to 10, we have gotten rid of a lot of the noise and still captured a large proportion of the variance. 

## Choosing number of archetypes

Fit to k = 2 to 8 to find the best number of archetypes. We will look at the variance explained by each archetype as well as the t-ratios. To choose a final number to move forward with the analysis, we will run a randomization test to get a p-value for each number of archetypes (t-ratio test). 
```{r}

library(ParetoTI)
library(cowplot)
library(ggplot2)
library(RColorBrewer)
library(reshape2)
library(factoextra)
library(ggfortify)
library(cluster)
# install.packages("data.table")
##################################
# load(file="../../out/cell-lines/ParetoTI/X_magic_for_ParetoTI.Robj")

x <- data[["RNA"]]@scale.data

x_pca <- read.csv('../../out/human-tumors/ParetoTI/pca_embedding.csv',header = TRUE,row.names = 1)
x_pca <- t(x_pca)

x_pca <- x_pca[1:12,] #keep only top 11 PCs
loadings<- read.csv('../../out/human-tumors/ParetoTI/pca_feature_loadings.csv', header = TRUE, row.names = 1)
loadings <- as.matrix(loadings)
```
``` {r}
arc_ks_8 = k_fit_pch(x_pca, ks = 2:8, check_installed = T,
                   bootstrap = T, bootstrap_N = 200, maxiter = 1000,
                   bootstrap_type = "m", seed = 2543, 
                   volume_ratio = "t_ratio", # set to "none" if too slow
                   delta=0, conv_crit = 1e-04, order_type = "align",
                   sample_prop = 0.75)

# Show variance explained by a polytope with each k (cumulative)
plot_arc_var(arc_ks_8, type = "varexpl", point_size = 2, line_size = 1.5) + theme_bw()
plot_arc_var(arc_ks_8, type = "res_varexpl", point_size = 2, line_size = 1.5) + theme_bw()
plot_arc_var(arc_ks_8, type = "total_var", point_size = 2, line_size = 1.5) +
  theme_bw() +
  ylab("Mean variance in position of vertices")
plot_arc_var(arc_ks_8, type = "t_ratio", point_size = 2, line_size = 1.5) + theme_bw()

```

``` {r}
arc_ks = k_fit_pch(x_pca, ks = 2:8, check_installed = T,
                   bootstrap = T, bootstrap_N = 200, maxiter = 1000,
                   bootstrap_type = "m", seed = 2543, 
                   volume_ratio = "t_ratio", # set to "none" if too slow
                   delta=0, conv_crit = 1e-04, order_type = "align",
                   sample_prop = 0.75)

# Show variance explained by a polytope with each k (cumulative)
plot_arc_var(arc_ks, type = "varexpl", point_size = 2, line_size = 1.5) + theme_bw()
plot_arc_var(arc_ks, type = "res_varexpl", point_size = 2, line_size = 1.5) + theme_bw()
plot_arc_var(arc_ks, type = "total_var", point_size = 2, line_size = 1.5) +
  theme_bw() +
  ylab("Mean variance in position of vertices")
plot_arc_var(arc_ks, type = "t_ratio", point_size = 2, line_size = 1.5) + theme_bw()

```

## T-ratio tests

```{r}
i = 6
arc <- fit_pch(x_pca, noc = i, delta = 0, conv_crit = 1e-04, maxiter = 500)

start = Sys.time()
pch_rand = randomise_fit_pch(x_pca, arc_data = arc,
                             n_rand = 1000,
                             replace = FALSE, bootstrap_N = NA,
                             volume_ratio = "t_ratio",
                             maxiter = 500, delta = 0, conv_crit = 1e-4,
                             type = "m", clust_options = list(cores = 3))
# use type m to run on a single machine or cloud
# type = "m", clust_options = list(cores = 3))
# use clustermq (type cmq) to run as jobs on a computing cluster (higher parallelisation)
# type = "cmq", clust_options = list(njobs = 10)) 


pdf(sprintf('../../figures/ParetoTI/human-tumors/%s_t-ratio_test.pdf', i))
plot.r_pch_fit(pch_rand, type = c("t_ratio"), nudge_y = 5)
dev.off()
  # This analysis took:
print(Sys.time() - start)

#  k  var_name   var_obs    p_value
# 1: 3   varexpl 0.8551638 0.23423423
# 2: 3   t_ratio 0.8051166 0.03003003
# 3: 3 total_var        NA        NaN

#   k  var_name      var_obs p_value
# 1: 4   varexpl 0.8552862213   0.958
# 2: 4   t_ratio 0.0000887529   0.999
# 3: 4 total_var           NA     NaN

#    k  var_name   var_obs p_value
# 1: 5   varexpl 0.9467342   0.001
# 2: 5   t_ratio 0.3798273   0.001
# 3: 5 total_var        NA     NaN
#    k  var_name   var_obs p_value
# 1: 6   varexpl 0.9624359   0.001
# 2: 6   t_ratio 0.3063444   0.001
# 3: 6 total_var        NA     NaN
```

# Fitting archetypes using PCHA

We'll start by fitting 6 archetypes and finding the enriched genes and gene sets for each.

``` {r}
arc <- fit_pch(x_pca, noc = 3, delta = 0, conv_crit = 1e-04, maxiter = 500)

##################################

# Fit 6 archetypes with bootstrapping for robustness 
arc_rob = fit_pch_bootstrap(x_pca, n = 200, sample_prop = .8, seed = 2543, delta = 1, noc = 3)
arc_ave <- average_pch_fits(arc_rob)
save(arc, file="../../out/human-tumors/ParetoTI/arc.Robj")
save(arc_rob, file="../../out/human-tumors/ParetoTI/arc_rob.Robj")
save(arc_ave, file="../../out/human-tumors/ParetoTI/arc_ave.Robj")
write.csv(arc$XC, file="../../out/human-tumors/ParetoTI/arc_positions_pca.csv")
write.csv(arc_ave$XC, file="../../out/human-tumors/ParetoTI/arc_ave_positions_pca.csv")

reconstruct_from_pca <- function(x, loadings, arc_XC, nComp = 12){
    mu = colMeans(t(x)) #mean for each feature (gene)
    # Xhat = [r(samples) x c(PCs)] %*% [r(PCs) x c(Features)] 
    # Xhat = [r(samples) x c(features)]
    Xhat = arc_XC[,1:nComp] %*% t(loadings[,1:nComp])
    Xhat = scale(Xhat, center = -mu, scale = FALSE)
    return(Xhat)}
x_recon <- reconstruct_from_pca(x, loadings, x_pca)
arc_genespace <- reconstruct_from_pca(x = x, loadings = loadings, arc_XC = t(arc_ave$XC))
write.csv(t(arc_genespace), '../../out/human-tumors/ParetoTI/arc_gene-space.csv')
##################################
# 
# load(file="../int/single-cell/arc.Robj")
# load( file="../int/single-cell/arc_rob.Robj")
# load( file="../int/single-cell/arc_ave.Robj")

Idents(object = data) <- data@meta.data$tumor #add cell line labels

cols <- c(brewer.pal(9, "Set1"),'gray')

plot_arc(arc_data = arc, data = x_pca,
                   which_dimensions = 1:2,colors = cols,
                  data_lab = as.character(Idents(data))) + theme_bw()
p_pca = plot_arc(arc_data = arc, data = x_pca, 
                 which_dimensions = 1:3, line_size = 1.5,
                 colors = cols,
                  data_lab = as.character(Idents(data)),
                 text_size = 60, data_size = 2) 
plotly::layout(p_pca, title = "Archetypes for Top 10 PCs")
htmlwidgets::saveWidget(p_pca, "../../figures/ParetoTI/human-tumors/robust_archetypes.html")

p_pca = plot_arc(arc_data = arc_ave, data = x_pca, 
                 which_dimensions = 1:3, line_size = 1.5,
                 data_lab = as.numeric(data[['RNA']]@scale.data['NEUROD2',]),
                 text_size = 60, data_size = 3) 
plotly::layout(p_pca, title = "Expression of NEUROD2 in PCA")
htmlwidgets::saveWidget(p_pca, "../../figures/ParetoTI/human-tumors/NEUROD2.HTML")

p_pca = plot_arc(arc_data = arc_ave, data = x_pca, 
                 which_dimensions = 1:3, line_size = 1.5,
                 data_lab = as.numeric(data[['RNA']]@data['YAP1',]),
                 text_size = 60, data_size = 3) 
plotly::layout(p_pca, title = "Expression of YAP1 in PCA")
htmlwidgets::saveWidget(p_pca, "../../figures/ParetoTI/human-tumors/YAP1.html")

# p_pca = plot_arc(arc_data = arc_ave, data = x_pca, 
#                  which_dimensions = 1:3, line_size = 1.5,
#                  data_lab = as.numeric(data[['RNA']]@data['POU2F3',]),
#                  text_size = 60, data_size = 3) 
# plotly::layout(p_pca, title = "Expression of POU2F3 in PCA")
# htmlwidgets::saveWidget(p_pca, "../../figures/ParetoTI/human-tumors/POU2F3.html")

p_pca = plot_arc(arc_data = arc_ave, data = x_pca, 
                 which_dimensions = 1:3, line_size = 1.5,
                 data_lab = as.numeric(data[['RNA']]@data['ASCL1',]),
                 text_size = 60, data_size = 3) 
plotly::layout(p_pca, title = "Expression of ASCL1 in PCA")
htmlwidgets::saveWidget(p_pca, "../../figures/ParetoTI/human-tumors/ASCL1.html")

```

# Determine enriched genes and gene sets to define archetypes

This will be especially interesting for comparing archetypes 4 and 5, since it is pretty clear that the shape of the data is not clearly defined without an archetype at 5, but not many cells actually lie close to it. We use the ParetoTI package to evaulate gene sets enriched in each location. 

```{r}
library(matrixStats)

# x.orig <- as.matrix(data[['RNA']]@data)
# 
# rownames(x.orig)<-gsub("-", ".", rownames(x.orig))
# 
# 
# 
# ix <- which(rownames(x.orig) %in%c('1.Sep','10.Sep','11.Mar','11.Sep','2.Sep','3.Mar','3.Sep','4.Mar','4.Sep','5.Sep', '5.Mar','6.Sep','6.Mar','7.Sep','7.Mar','8.Sep','8.Mar','9.Sep','9.Mar','RP11-206L10.1'))
# ix <- which(rownames(x.orig) %in% c('7SK.1'))
# clean <- x.orig[-ix, ]


activ_pi = measure_activity(as.matrix(data[['RNA']]@scale.data), activity_method = 'pseudoinverse',# row names are assumed to be gene identifiers,
                         which = 'BP', return_as_matrix = F,
                         taxonomy_id = 9606, keytype = "ALIAS", #9606 is human, 10090 is mouse.
                         lower = 10, upper = 1000)
                         # aucell_options =list(aucMaxRank =
                          # nrow(as.matrix(x.orig)) * 0.05, binary = F, nCores = 3, plotStats = TRUE))
save(activ_pi, file="../../out/cell-lines/ParetoTI/activ_pi.Robj")

# activ_pi <- within(activ_pi, rm('2__deoxyribonucleotide_biosynthetic_process','2__deoxyribonucleotide_metabolic_process','2_oxoglutarate_metabolic_process','3__phosphoadenosine_5__phosphosulfate_metabolic_process',
# '3__UTR_mediated_mRNA_destabilization',
# '3__UTR_mediated_mRNA_stabilization',
# '7_methylguanosine_mRNA_capping',
# '7_methylguanosine_RNA_capping',
# '4_hydroxyproline_metabolic_process'))
# 
# activ_pi <- within(activ_pi, rm(`_de_novo__posttranslational_protein_folding`,
# `_de_novo__protein_folding`,
# `poly_A_+_mRNA_export_from_nucleus`))


data_attr = merge_arch_dist(arc_data = arc_ave, data = x_pca, 
                            feature_data = as.matrix(data[['RNA']]@scale.data),
                            colData = activ_pi, 
                            dist_metric = c("euclidean", "arch_weights")[1],
                            colData_id = "cells", rank = F) 
save(data_attr, file="../../out/cell-lines/ParetoTI/data_attr.Robj")



enriched_genes = find_decreasing_wilcox(data_attr$data, data_attr$arc_col,
                                features = data_attr$features_col,
                                bin_prop = 0.05, method = "BioQC")
write.csv(enriched_genes, '../../out/cell-lines/ParetoTI/enriched-genes.csv')

enriched_sets = find_decreasing_wilcox(data_attr$data, data_attr$arc_col,
                                features = data_attr$colData_col,
                                bin_prop = 0.05, method = "BioQC")
write.csv(enriched_sets, '../../out/cell-lines/ParetoTI/enriched_sets.csv')

labs = get_top_decreasing(summary_genes = enriched_genes, summary_sets = enriched_sets,
                          cutoff_genes = 0.05,cutoff_sets = 0.05,
                          cutoff_metric = "wilcoxon_p_val", 
                          p.adjust.method = "fdr", 
                          order_by = "mean_diff", order_decreasing = T)
# 1. N (H524)
# 2. A/N (CORL279)
# 3. A2 (DMS53)
# 4. Y (H841)
# 5. A (H69)
# 6. A2 (DMS454)

# enriched_genes_gam = find_decreasing(data_attr$data, data_attr$arc_col,
#                                 features = data_attr$features_col, return_only_summary = TRUE)
# write.csv(enriched_genes_gam, '../data/single-cell/enriched_genes_gam.csv')
# 
# enriched_sets_gam = find_decreasing(data_attr$data, data_attr$arc_col,
#                                 features = data_attr$colData_col, return_only_summary = TRUE)
# write.csv(enriched_sets_gam, '../data/single-cell/enriched_sets_gam.csv')
# 
# 
# 
# labs_gam = get_top_decreasing(summary_genes = enriched_genes_gam, summary_sets = enriched_sets_gam,
#                           cutoff_genes = 0.05,cutoff_sets = 0.05,
#                           cutoff_metric = "mean_prob", 
#                           p.adjust.method = "none", 
#                           order_by = "deriv50", order_decreasing = F,
#                           min_max_diff_cutoff_g = .05)
# 
# fit_arc_gam_1('ASCL1', "archetype_2", data_attr)
```

## Save enrichments to be compared to bulk archetypes

The format we need the enrichment file in is a csv with the columns:
archetype #,Feature Name,P value (Mann-Whitney),Median Difference,Mean Difference,Significant after Benjamini-Hochberg correction?,Is first bin maximal?

```{r save-enriched}
# "x_name" --> "archetype #"
# "y_name" --> "Feature Name"
# "mean_diff" --> "Mean Difference"
# "p" --> "P value (Mann-Whitney)"
# subset to p < 0.05
# Feature names need to be capitalized and remove underscores
enriched_sets <- enriched_sets[enriched_sets$p < 0.05, ]
library(dplyr)
enriched_sets <- enriched_sets %>%
  rename("archetype #" = x_name,
         "Feature Name" = y_name,
         "Mean Difference" = mean_diff,
         "P value (Mann-Whitney)" = p)
enriched_sets$`Feature Name` <- gsub("_", " ", enriched_sets$`Feature Name`)
enriched_sets$`Feature Name` <- toupper(enriched_sets$`Feature Name`)

write.csv(enriched_sets, '../../out/cell-lines/ParetoTI/single-cell-continuous_significant.csv')

```

```{r session-info}
session_info. = devtools::session_info()
session_info.
```

