---
title: "Comparing Original Cell Line Archetypes with Single Cell Cell Line Archetypes"
author: "Sarah Groves"
date: "12/14/2021"
output: 
    html_notebook:
        self_contained: yes
        toc: yes
        toc_depth: 4
---

This document is based on characterizeArchetypes.R from Hausser Code from Hausser et al. It compares the original archetypes (5) found on cell line data to the combined archetypes found on all the data by figuring out which archetypes have the most overlapping gene ontology enrichments.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# BiocManager::install(c('fdrtool','ade4','rgl'))
library(fdrtool)
library(ade4)
# library(rgl)
library(ggrepel)
library(stringr)
# BiocManager::install('tidyverse')
library(tidyverse)
library(forcats)
```

```{r}
##################################################
## Make matrix of bulk archetypes vs single cell archetypes

# Read in significant enrichment for bulk archetypes
SAMSig <- read.csv("../ParTI-code/human-cell-lines/out/5/322234/322234_continuous_significant.csv")
SAMSig <- SAMSig %>% dplyr::filter(Mean.Difference > 0.1)


# # This makes a list of all of the archetypes (# and p value) for each dataset, and makes a list of files, one enrichment file per dataset. This is unnecessary for us to do
# cancerIDs <-
#     read_tsv("../TCGA_frac_nArchs.tab", col_names=F) %>% .[,1] %>%
#     unlist %>% as.character %>%
#     setdiff(c("HNSC", "LUAD", "BRCA"))
# arcsMSigFiles <-
#     c(paste("../", cancerIDs,
#             "_UCSC/MSigDBenrichment_continuous_significant.csv",
#             sep=""),
#       "~/work/cancerTaskAtlas/brca_metabric/MSigDBenrichment_continuous_significant.csv")
# cancerIDs <- c(cancerIDs, "BRCA")
# cbind(cancerIDs, arcsMSigFiles)


arcsMSig <-read.csv('../../data/single-cell-rna-seq/PNECs/single-cell-continuous_significant.csv') %>%
            dplyr::filter(Mean.Difference > 0.1)
vector <-arcsMSig$Feature.Name
vector2 <- arcsMSig$Feature.Name

for(i in 1:length(vector)){
  vector2[i] <- paste("GO",vector[i])
}

arcsMSig$Feature.Name <- vector2

featsUniv <- dplyr::union(SAMSig %>% dplyr::select("Feature.Name"),
                   arcsMSig %>% dplyr::select("Feature.Name"))

arcIdx <- 1;
SAmapping <- # iterate over tissue archetypes
            map(unlist(arcsMSig %>% dplyr::select(archetype..) %>% unique), function(arcIdx) {
                SAidx <- 1;
                arcScores <- # iterate over bulk-archetypes
                    map(unlist(SAMSig %>% dplyr::select(archetype..) %>% unique),
                        function(SAidx) {
                            SAfeats <- SAMSig %>%
                                dplyr::filter(archetype.. == SAidx) %>%
                                dplyr::select(Feature.Name)
                            arcFeats <- arcsMSig %>%
                                dplyr::filter(archetype.. == arcIdx) %>%
                                dplyr::select(Feature.Name)
                            ## nrow(SAfeats)
                            ## nrow(arcFeats)
                            
                            expIntersect <- nrow(SAfeats) * nrow(arcFeats) / nrow(featsUniv)
                            ## phyper(q, m, n, k, lower.tail = TRUE, log.p = FALSE)
                            ## q,x: number of white balls drawn without
                            ## replacement from an urn which contains
                            ## both black and white balls.
                            ## m: the number of white balls in the urn.
                            ## n: the number of black balls in the urn.
                            ## k: the number of balls drawn from the urn.
                            p <- phyper(q=intersect(SAfeats, arcFeats) %>% nrow,
                                        m=SAfeats %>% nrow,
                                        n=nrow(featsUniv) - nrow(SAfeats),
                                        k=arcFeats %>% nrow,
                                        lower.tail=F)
                            foldEnrich <- nrow(intersect(SAfeats, arcFeats)) / expIntersect;

                            ## Bonferroni
                            cutOff <- .01 /
                                ((arcsMSig %>% dplyr::select(archetype..) %>%
                                  unique %>% nrow) *
                                 (SAMSig %>% dplyr::select(archetype..) %>%
                                  unique %>% nrow))

                            return(c("obs"=nrow(intersect(SAfeats, arcFeats)),
                                     "exp"=expIntersect,
                                     "p"=p,
                                     "isSignif"=p<cutOff))
                        })
                arcScoresT <-
                    sapply(arcScores, function(x) { x }) %>% t %>% as.data.frame %>% 
                    rownames_to_column() %>%
                    mutate(tissueArch=arcIdx) %>%
                    rename("univArch"=rowname)
                return(arcScoresT)
            })
save.image("~/Documents/Github/Groves-CellSys2021/notebooks/ParTI-code/SAmapping.RData")
# load("/Users/smgroves/Documents/Github/Groves-CellSys2021/notebooks/ParTI-code/SAmapping.RData")
```