---
title: "Single Cell SCLC Cell Line Archetypes"
output:
  html_document:
    df_print: paged
---
The purpose of the notebook is to apply PCHA to the single cell cell line data. 

```{r include = FALSE}
library(reticulate)
reticulate::use_condaenv("/Users/smgroves/Documents/anaconda3/envs/mazebox_env", conda = "auto", required = TRUE) # set TRUE to force R to use reticulate_PCHA
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")


```


# Archetypal Analysis

We will now apply the archetype analysis to the single cell data. Since we've reduced the PCs to 10, we have gotten rid of a lot of the noise and still captured a large proportion of the variance. 

## Choosing number of archetypes

Fit to k = 2 to 8 to find the best number of archetypes. We will look at the variance explained by each archetype as well as the t-ratios. To choose a final number to move forward with the analysis, we will run a randomization test to get a p-value for each number of archetypes (t-ratio test). 
```{r}

library(ParetoTI)
library(cowplot)
library(ggplot2)
library(RColorBrewer)
library(reshape2)
library(factoextra)
library(ggfortify)
library(cluster)
# install.packages("data.table")
##################################
# load(file="../../out/cell-lines/ParetoTI/X_magic_for_ParetoTI.Robj")
top_pcs = 11

x_pca <- read.csv('../../data/single-cell-rna-seq/PNECs/magic_pca.csv',header = TRUE,row.names = 1)
x_pca <- t(x_pca)

# x_pca <- x_pca[1:top_pcs,] #keep only top 11 PCs

```

``` {r}
arc_ks = k_fit_pch(x_pca, ks = 2:8, check_installed = T,
                   bootstrap = T, bootstrap_N = 200, maxiter = 1000,
                   bootstrap_type = "m", seed = 2543, 
                   volume_ratio = "t_ratio", # set to "none" if too slow
                   delta=0, conv_crit = 1e-04, order_type = "align",
                   sample_prop = 0.75)

# Show variance explained by a polytope with each k (cumulative)
plot_arc_var(arc_ks, type = "varexpl", point_size = 2, line_size = 1.5) + theme_bw()
plot_arc_var(arc_ks, type = "res_varexpl", point_size = 2, line_size = 1.5) + theme_bw()
plot_arc_var(arc_ks, type = "total_var", point_size = 2, line_size = 1.5) +
  theme_bw() +
  ylab("Mean variance in position of vertices")
plot_arc_var(arc_ks, type = "t_ratio", point_size = 2, line_size = 1.5) + theme_bw()

```


## T-ratio tests

```{r}
i = 5
arc <- fit_pch(x_pca, noc = i, delta = 0, conv_crit = 1e-04, maxiter = 500)

start = Sys.time()
pch_rand = randomise_fit_pch(x_pca, arc_data = arc,
                             n_rand = 1000,
                             replace = FALSE, bootstrap_N = NA,
                             volume_ratio = "t_ratio",
                             maxiter = 500, delta = 0, conv_crit = 1e-4,
                             type = "m", clust_options = list(cores = 3))
# use type m to run on a single machine or cloud
# type = "m", clust_options = list(cores = 3))
# use clustermq (type cmq) to run as jobs on a computing cluster (higher parallelisation)
# type = "cmq", clust_options = list(njobs = 10)) 


pdf(sprintf('./figures/%s_t-ratio_test.pdf', i))
plot.r_pch_fit(pch_rand, type = c("t_ratio"), nudge_y = 5)
dev.off()
  # This analysis took:
print(Sys.time() - start)
#    k  var_name   var_obs     p_value
# 1: 3   varexpl 0.8944519 0.001006036
# 2: 3   t_ratio 0.6004021 0.001006036
# 3: 3 total_var        NA         NaN
#    k  var_name   var_obs     p_value
# 1: 4   varexpl 0.9339312 0.001003009
# 2: 4   t_ratio 0.2975490 0.001003009
# 3: 4 total_var        NA         NaN
#   k  var_name   var_obs     p_value
# 1: 5   varexpl 0.9695800 0.001007049
# 2: 5   t_ratio 0.2128127 0.001007049
# 3: 5 total_var        NA         NaN

```

# Fitting archetypes using PCHA

We'll start by fitting 3 archetypes and finding the enriched genes and gene sets for each.

``` {r}
arc <- fit_pch(x_pca, noc = 3, delta = 0, conv_crit = 1e-04, maxiter = 500)

##################################

# Fit 4 archetypes with bootstrapping for robustness
arc_rob = fit_pch_bootstrap(x_pca, n = 200, sample_prop = .8, seed = 2543, delta = 1,
                            noc = 3)
arc_ave <- average_pch_fits(arc_rob)
# save(arc, file="../../out/cell-lines/ParetoTI/arc.Robj")
# save(arc_rob, file="../../out/cell-lines/ParetoTI/arc_rob.Robj")
# save(arc_ave, file="../../out/cell-lines/ParetoTI/arc_ave.Robj")
# write.csv(arc$XC, file="../../out/cell-lines/ParetoTI/arc_positions_pca.csv")
# write.csv(arc_ave$XC, file="../../out/cell-lines/ParetoTI/arc_ave_positions_pca.csv")
# 
# reconstruct_from_pca <- function(x, loadings, arc_XC, nComp = 11){
#     mu = colMeans(t(x)) #mean for each feature (gene)
#     # Xhat = [r(samples) x c(PCs)] %*% [r(PCs) x c(Features)] 
#     # Xhat = [r(samples) x c(features)]
#     Xhat = arc_XC[,1:nComp] %*% t(loadings[,1:nComp])
#     Xhat = scale(Xhat, center = -mu, scale = FALSE)
#     return(Xhat)}
# x_recon <- reconstruct_from_pca(x, loadings, x_pca)
# arc_genespace <- reconstruct_from_pca(x = x, loadings = loadings, arc_XC = t(arc_ave$XC))
# write.csv(t(arc_genespace), '../../out/cell-lines/ParetoTI/arc_gene-space.csv')
# 
# arc_genespace <- reconstruct_from_pca(x = x, loadings = loadings, arc_XC = t(arc$XC))
# write.csv(t(arc_genespace), '../../out/cell-lines/ParetoTI/arc_single-fit-gene-space.csv')
##################################
# 
# load(file="../int/single-cell/arc.Robj")
# load( file="../int/single-cell/arc_rob.Robj")

```

```{r}
plot_arc(arc_data = arc, data = x_pca,
                   which_dimensions = 1:2) + theme_bw()
```

```{r}
arc_5 <- fit_pch(x_pca, noc = 5, delta = 0, conv_crit = 1e-04, maxiter = 500)

plot_arc(arc_data = arc_5, data = x_pca) + theme_bw()

```

# Determine enriched genes and gene sets to define archetypes

This will be especially interesting for comparing archetypes 4 and 5, since it is pretty clear that the shape of the data is not clearly defined without an archetype at 5, but not many cells actually lie close to it. We use the ParetoTI package to evaulate gene sets enriched in each location. 

```{r}
library(matrixStats)

# x.orig <- as.matrix(data[['RNA']]@data)
# 
# rownames(x.orig)<-gsub("-", ".", rownames(x.orig))
# 
# 
# 
# ix <- which(rownames(x.orig) %in%c('1.Sep','10.Sep','11.Mar','11.Sep','2.Sep','3.Mar','3.Sep','4.Mar','4.Sep','5.Sep', '5.Mar','6.Sep','6.Mar','7.Sep','7.Mar','8.Sep','8.Mar','9.Sep','9.Mar','RP11-206L10.1'))
# ix <- which(rownames(x.orig) %in% c('7SK.1'))
# clean <- x.orig[-ix, ]
data =  read.csv('../../data/single-cell-rna-seq/PNECs/log_data.csv',header = TRUE,row.names = 1)

activ_pi = measure_activity(as.matrix(t(data)), activity_method = 'pseudoinverse',# row names are assumed to be gene identifiers,
                         which = 'BP', return_as_matrix = F,
                         taxonomy_id = 10090, keytype = "ALIAS", #9606 is human, 10090 is mouse.
                         lower = 10, upper = 1000)
                         # aucell_options =list(aucMaxRank =
                          # nrow(as.matrix(x.orig)) * 0.05, binary = F, nCores = 3, plotStats = TRUE))
save(activ_pi, file="../../data/single-cell-rna-seq/PNECs/activ_pi.Robj")

# activ_pi <- within(activ_pi, rm('2__deoxyribonucleotide_biosynthetic_process','2__deoxyribonucleotide_metabolic_process','2_oxoglutarate_metabolic_process','3__phosphoadenosine_5__phosphosulfate_metabolic_process',
# '3__UTR_mediated_mRNA_destabilization',
# '3__UTR_mediated_mRNA_stabilization',
# '7_methylguanosine_mRNA_capping',
# '7_methylguanosine_RNA_capping',
# '4_hydroxyproline_metabolic_process'))
# 
# activ_pi <- within(activ_pi, rm(`_de_novo__posttranslational_protein_folding`,
# `_de_novo__protein_folding`,
# `poly_A_+_mRNA_export_from_nucleus`))


data_attr = merge_arch_dist(arc_data = arc_ave, data = x_pca, 
                            feature_data = as.matrix(t(data)),
                            colData = activ_pi, 
                            dist_metric = c("euclidean", "arch_weights")[1],
                            colData_id = "cells", rank = F) 
save(data_attr, file="../../data/single-cell-rna-seq/PNECs/data_attr.Robj")



enriched_genes = find_decreasing_wilcox(data_attr$data, data_attr$arc_col,
                                features = data_attr$features_col,
                                bin_prop = 0.05, method = "BioQC")
write.csv(enriched_genes, '../../data/single-cell-rna-seq/PNECs/enriched-genes.csv')

enriched_sets = find_decreasing_wilcox(data_attr$data, data_attr$arc_col,
                                features = data_attr$colData_col,
                                bin_prop = 0.05, method = "BioQC")
write.csv(enriched_sets, '../../data/single-cell-rna-seq/PNECs/enriched_sets.csv')

labs = get_top_decreasing(summary_genes = enriched_genes, summary_sets = enriched_sets,
                          cutoff_genes = 0.05,cutoff_sets = 0.05,
                          cutoff_metric = "wilcoxon_p_val", 
                          p.adjust.method = "fdr", 
                          order_by = "mean_diff", order_decreasing = T)

# enriched_genes_gam = find_decreasing(data_attr$data, data_attr$arc_col,
#                                 features = data_attr$features_col, return_only_summary = TRUE)
# write.csv(enriched_genes_gam, '../data/single-cell/enriched_genes_gam.csv')
# 
# enriched_sets_gam = find_decreasing(data_attr$data, data_attr$arc_col,
#                                 features = data_attr$colData_col, return_only_summary = TRUE)
# write.csv(enriched_sets_gam, '../data/single-cell/enriched_sets_gam.csv')
# 
# 
# 
# labs_gam = get_top_decreasing(summary_genes = enriched_genes_gam, summary_sets = enriched_sets_gam,
#                           cutoff_genes = 0.05,cutoff_sets = 0.05,
#                           cutoff_metric = "mean_prob", 
#                           p.adjust.method = "none", 
#                           order_by = "deriv50", order_decreasing = F,
#                           min_max_diff_cutoff_g = .05)
# 
# fit_arc_gam_1('ASCL1', "archetype_2", data_attr)
```

## Save enrichments to be compared to bulk archetypes

The format we need the enrichment file in is a csv with the columns:
archetype #,Feature Name,P value (Mann-Whitney),Median Difference,Mean Difference,Significant after Benjamini-Hochberg correction?,Is first bin maximal?

```{r save-enriched}
# "x_name" --> "archetype #"
# "y_name" --> "Feature Name"
# "mean_diff" --> "Mean Difference"
# "p" --> "P value (Mann-Whitney)"
# subset to p < 0.05
# Feature names need to be capitalized and remove underscores
enriched_sets <- enriched_sets[enriched_sets$p < 0.05, ]
library(dplyr)
enriched_sets <- enriched_sets %>%
  rename("archetype #" = x_name,
         "Feature Name" = y_name,
         "Mean Difference" = mean_diff,
         "P value (Mann-Whitney)" = p)
enriched_sets$`Feature Name` <- gsub("_", " ", enriched_sets$`Feature Name`)
enriched_sets$`Feature Name` <- toupper(enriched_sets$`Feature Name`)

write.csv(enriched_sets, '../../data/single-cell-rna-seq/PNECs/single-cell-continuous_significant.csv')

```

```{r}


data_attr5 = merge_arch_dist(arc_data = arc_5, data = x_pca, 
                            feature_data = as.matrix(t(data)),
                            colData = activ_pi, 
                            dist_metric = c("euclidean", "arch_weights")[1],
                            colData_id = "cells", rank = F) 
save(data_attr, file="../../data/single-cell-rna-seq/PNECs/data_attr5.Robj")



enriched_genes = find_decreasing_wilcox(data_attr5$data, data_attr5$arc_col,
                                features = data_attr5$features_col,
                                bin_prop = 0.05, method = "BioQC")
write.csv(enriched_genes, '../../data/single-cell-rna-seq/PNECs/enriched-genes5.csv')

enriched_sets = find_decreasing_wilcox(data_attr5$data, data_attr5$arc_col,
                                features = data_attr5$colData_col,
                                bin_prop = 0.05, method = "BioQC")
write.csv(enriched_sets, '../../data/single-cell-rna-seq/PNECs/enriched_sets5.csv')

labs = get_top_decreasing(summary_genes = enriched_genes, summary_sets = enriched_sets,
                          cutoff_genes = 0.05,cutoff_sets = 0.05,
                          cutoff_metric = "wilcoxon_p_val", 
                          p.adjust.method = "fdr", 
                          order_by = "mean_diff", order_decreasing = T)

```
# Plasschaert data

```{r include = FALSE}
library(reticulate)
reticulate::use_condaenv("/Users/smgroves/Documents/anaconda3/envs/mazebox_env", conda = "auto", required = TRUE) # set TRUE to force R to use reticulate_PCHA
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")


```


# Archetypal Analysis

We will now apply the archetype analysis to the single cell data. Since we've reduced the PCs to 10, we have gotten rid of a lot of the noise and still captured a large proportion of the variance. 

## Choosing number of archetypes

Fit to k = 2 to 8 to find the best number of archetypes. We will look at the variance explained by each archetype as well as the t-ratios. To choose a final number to move forward with the analysis, we will run a randomization test to get a p-value for each number of archetypes (t-ratio test). 
```{r}

library(ParetoTI)
library(cowplot)
library(ggplot2)
library(RColorBrewer)
library(reshape2)
library(factoextra)
library(ggfortify)
library(cluster)
# install.packages("data.table")
##################################
# load(file="../../out/cell-lines/ParetoTI/X_magic_for_ParetoTI.Robj")
top_pcs = 11

x_pca <- read.csv('../../data/single-cell-rna-seq/PNECs/magic_pca_plasschaert.csv',header = TRUE,row.names = 1)
x_pca <- t(x_pca)

# x_pca <- x_pca[1:top_pcs,] #keep only top 11 PCs

```

``` {r}
arc_ks = k_fit_pch(x_pca, ks = 2:8, check_installed = T,
                   bootstrap = T, bootstrap_N = 200, maxiter = 1000,
                   bootstrap_type = "m", seed = 2543, 
                   volume_ratio = "t_ratio", # set to "none" if too slow
                   delta=0, conv_crit = 1e-04, order_type = "align",
                   sample_prop = 0.75)

# Show variance explained by a polytope with each k (cumulative)
plot_arc_var(arc_ks, type = "varexpl", point_size = 2, line_size = 1.5) + theme_bw()
plot_arc_var(arc_ks, type = "res_varexpl", point_size = 2, line_size = 1.5) + theme_bw()
plot_arc_var(arc_ks, type = "total_var", point_size = 2, line_size = 1.5) +
  theme_bw() +
  ylab("Mean variance in position of vertices")
plot_arc_var(arc_ks, type = "t_ratio", point_size = 2, line_size = 1.5) + theme_bw()

```
```{r}
i = 5
arc <- fit_pch(x_pca, noc = i, delta = 0, conv_crit = 1e-04, maxiter = 500)

start = Sys.time()
pch_rand = randomise_fit_pch(x_pca, arc_data = arc,
                             n_rand = 1000,
                             replace = FALSE, bootstrap_N = NA,
                             volume_ratio = "t_ratio",
                             maxiter = 500, delta = 0, conv_crit = 1e-4,
                             type = "m", clust_options = list(cores = 3))
# use type m to run on a single machine or cloud
# type = "m", clust_options = list(cores = 3))
# use clustermq (type cmq) to run as jobs on a computing cluster (higher parallelisation)
# type = "cmq", clust_options = list(njobs = 10)) 


pdf(sprintf('./figures/%s_t-ratio_test.pdf', i))
plot.r_pch_fit(pch_rand, type = c("t_ratio"), nudge_y = 5)
dev.off()
  # This analysis took:
print(Sys.time() - start)
# 
#    k  var_name   var_obs     p_value
# 1: 3   varexpl 0.6457321 0.072597137
# 2: 3   t_ratio 0.5041697 0.001022495
# 3: 3 total_var        NA         NaN
#    k  var_name   var_obs     p_value
# 1: 4   varexpl 0.6871283 0.068825911
# 2: 4   t_ratio 0.2122091 0.006072874
# 3: 4 total_var        NA         NaN
#    k  var_name    var_obs     p_value
# 1: 5   varexpl 0.72132958 0.012170385
# 2: 5   t_ratio 0.09463487 0.001014199
# 3: 5 total_var         NA         NaN

```

# Fitting archetypes using PCHA

We'll start by fitting 3 archetypes and finding the enriched genes and gene sets for each.

``` {r}
arc <- fit_pch(x_pca, noc = 3, delta = 0, conv_crit = 1e-04, maxiter = 500)

##################################

# Fit 4 archetypes with bootstrapping for robustness
arc_rob = fit_pch_bootstrap(x_pca, n = 200, sample_prop = .8, seed = 2543, delta = 1,
                            noc = 3)
arc_ave <- average_pch_fits(arc_rob)
# save(arc, file="../../out/cell-lines/ParetoTI/arc.Robj")
# save(arc_rob, file="../../out/cell-lines/ParetoTI/arc_rob.Robj")
# save(arc_ave, file="../../out/cell-lines/ParetoTI/arc_ave.Robj")
# write.csv(arc$XC, file="../../out/cell-lines/ParetoTI/arc_positions_pca.csv")
# write.csv(arc_ave$XC, file="../../out/cell-lines/ParetoTI/arc_ave_positions_pca.csv")
# 
# reconstruct_from_pca <- function(x, loadings, arc_XC, nComp = 11){
#     mu = colMeans(t(x)) #mean for each feature (gene)
#     # Xhat = [r(samples) x c(PCs)] %*% [r(PCs) x c(Features)] 
#     # Xhat = [r(samples) x c(features)]
#     Xhat = arc_XC[,1:nComp] %*% t(loadings[,1:nComp])
#     Xhat = scale(Xhat, center = -mu, scale = FALSE)
#     return(Xhat)}
# x_recon <- reconstruct_from_pca(x, loadings, x_pca)
# arc_genespace <- reconstruct_from_pca(x = x, loadings = loadings, arc_XC = t(arc_ave$XC))
# write.csv(t(arc_genespace), '../../out/cell-lines/ParetoTI/arc_gene-space.csv')
# 
# arc_genespace <- reconstruct_from_pca(x = x, loadings = loadings, arc_XC = t(arc$XC))
# write.csv(t(arc_genespace), '../../out/cell-lines/ParetoTI/arc_single-fit-gene-space.csv')
##################################
# 
# load(file="../int/single-cell/arc.Robj")
# load( file="../int/single-cell/arc_rob.Robj")

```

```{r}
plot_arc(arc_data = arc, data = x_pca,
                   which_dimensions = 1:2) + theme_bw()
```


```{r session-info}
session_info. = devtools::session_info()
session_info.
```

